plugins/protocols/session-lock.cpp

```diff
@@ -0,0 +1,435 @@
+            wf::scene::wlr_surface_node_t(lock_surface->surface, true /* autocommit */),
+            lock_surface(lock_surface),
+            output(output),
+            interaction(std::make_unique<lock_surface_keyboard_interaction>(lock_surface->surface,
+                lock_surface->output))
+        {}
+
+        void display()
+        {
+            auto layer_node = output->node_for_layer(wf::scene::layer::LOCK);
+            wf::scene::add_front(layer_node, shared_from_this());
+            wf::wlr_surface_controller_t::create_controller(lock_surface->surface, layer_node);
+            wf::get_core().seat->set_active_node(shared_from_this());
+            wf::get_core().seat->refocus();
+        }
+
+        void destroy()
+        {
+            wf::wlr_surface_controller_t::try_free_controller(this->lock_surface->surface);
+            wf::scene::remove_child(shared_from_this());
+            const char *name = this->output->handle ? this->output->handle->name : "(deleted)";
+            this->interaction = std::make_unique<wf::keyboard_interaction_t>();
+            LOGC(LSHELL, "lock_surface on ", name, " destroyed");
+        }
+
+        wf::keyboard_focus_node_t keyboard_refocus(wf::output_t *output) override
+        {
+            if (output != this->output)
+            {
+                return wf::keyboard_focus_node_t{};
+            }
+
+            wf::keyboard_focus_node_t node = {
+                .node = this,
+                .importance = wf::focus_importance::HIGH,
+                .allow_focus_below = false,
+            };
+            return node;
+        }
+
+        class lock_surface_keyboard_interaction : public wf::keyboard_interaction_t
+        {
+          public:
+            lock_surface_keyboard_interaction(wlr_surface *surface, wlr_output *output) :
+                surface(surface), output(output)
+            {}
+
+            void handle_keyboard_enter(wf::seat_t *seat)
+            {
+                wlr_seat_keyboard_enter(seat->seat, surface, nullptr, 0, nullptr);
+            }
+
+            void handle_keyboard_leave(wf::seat_t *seat)
+            {
+                wlr_seat_keyboard_clear_focus(seat->seat);
+            }
+
+            void handle_keyboard_key(wf::seat_t *seat, wlr_keyboard_key_event event)
+            {
+                wlr_seat_keyboard_notify_key(seat->seat, event.time_msec, event.keycode, event.state);
+            }
+
+          private:
+            wlr_surface *surface;
+            wlr_output *output;
+        };
+
+        wf::keyboard_interaction_t& keyboard_interaction()
+        {
+            return *interaction;
+        }
+
+      private:
+        wlr_session_lock_surface_v1 *lock_surface;
+        wf::output_t *output;
+        std::unique_ptr<wf::keyboard_interaction_t> interaction;
+    };
+
+    class backup_node : public simple_text_node_t
+    {
+      public:
```

Comment:
This node doesn't have keyboard or pointer interaction set - but it should have it, otherwise, it is not really 'locking' the desktop.

Also on my machine there are a few thin semi-transparent strips along some of the emoji's sides, I am not sure why that happens - I'm mentioning this because through them, I could see that I am clicking on the terminal behind the crashed swaylock :)

Reply:
Urg, yes. If I make the backup nodes transparent, it's pretty clear that they're not actually preventing keyboard or mouse input. Fixed in latest patch series (I think) by overriding find_node at. Not sure if this is correct or whether it will fix touch as well (I don't have a way to test touch).

Reply:
> Urg, yes. If I make the backup nodes transparent, it's pretty clear that they're not actually preventing keyboard or mouse input. Fixed in latest patch series (I think) by overriding find_node_at. Not sure if this is correct or whether it will fix touch as well (I don't have a way to test touch).

Yes, this is enough to fix touch, tablet and pointer. For the keyboard you'd also have to implement the refocus_keyboard() function very similarly to the lock surface node type.

Reply:
Ah, you already have this, nice! I'll test on multi-monitor as well to make sure the logic doesn't break somewhere and merge.